src/mpid/ch4/shm/xpmem/xpmem_recv.h

```diff
@@ -9,72 +9,316 @@
+      description : >-
+        Chunk size is used in XPMEM COOP P2P and determines the concurrency of
+        copy and overhead of atomic per chunk copy. Best chunk size should
+        balance these two factors.
+
+    - name        : MPIR_CVAR_CH4_XPMEM_COOP_COPY_THRESHOLD
+      category    : CH4
+      type        : int
+      default     : 32768
+      class       : device
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        When data size is larger than this threshold, cooperative copy is used;
+        if not, we need to fall back to single-side XPMEM copy.
+
+=== END_MPI_T_CVAR_INFO_BLOCK ===
+*/
+
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_do_lmt_coop_copy(const void *src_buf,
+                                                          size_t data_sz, void *dest_buf,
+                                                          OPA_int_t * counter_ptr,
+                                                          uint64_t req_ptr,
+                                                          MPIR_Request * local_req, int *fin_type,
+                                                          int *copy_type)
+{
+    int mpi_errno = MPI_SUCCESS;
+    size_t copy_sz;
+    int cur_chunk = 0, total_chunk = 0, num_local_copy = 0;
+    uint64_t cur_offset;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_DO_LMT_COOP_COPY);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_DO_LMT_COOP_COPY);
+
+    total_chunk =
+        data_sz / MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE +
+        (data_sz % MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE == 0 ? 0 : 1);
+
+    MPIR_Assert(total_chunk > 1);
+
+    /* TODO: implement OPA_fetch_and_incr_int uint64_t to directly change cur_offset
+     * instead of computing it in each copy */
+    while ((cur_chunk = OPA_fetch_and_incr_int(counter_ptr)) < total_chunk) {
+        cur_offset = ((uint64_t) cur_chunk) * MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE;
+        copy_sz =
+            cur_offset + MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE <=
+            data_sz ? MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE : data_sz - cur_offset;
+        mpi_errno =
+            MPIR_Localcopy(((char *) src_buf + cur_offset), copy_sz, MPI_BYTE,
+                           ((char *) dest_buf + cur_offset), copy_sz, MPI_BYTE);
+        MPIR_ERR_CHECK(mpi_errno);
+        num_local_copy++;
+    }
+
+    /* - If cur_chunk == total_chunk, copy is only locally complete (fin_type = LOCAL_FIN)
+     * - If cur_chunk == total_chunk + 1, copy is done by both sides (fin_type = BOTH_FIN)
+     * - If num_local_copy == total_chunk, it means the process copies all chunks (copy_type = MPIDI_XPMEM_COPY_ALL)
+     * - If num_local_copy == 0, it means the process copies zero chunk (copy_type = MPIDI_XPMEM_COPY_ZERO)
+     * - In other cases, it means both processes copy a part of chunks (copy_type = MPIDI_XPMEM_COPY_MIX) */
+    if (cur_chunk == total_chunk)
```

Comment:
Maybe it is simpler if we move the comment in each code line since most values are self-explained. E.g.,
```
    if (cur_chunk == total_chunk)
        *fin_type = MPIDI_XPMEM_LOCAL_FIN;
    else
        *fin_type = MPIDI_XPMEM_BOTH_FIN;  /* cur_chunk == total_chunk + 1, copy is done on both sides */

    if (num_local_copy == total_chunk)
        *copy_type = MPIDI_XPMEM_COPY_ALL;
    else if (num_local_copy == 0)
        *copy_type = MPIDI_XPMEM_COPY_ZERO;
    else
        *copy_type = MPIDI_XPMEM_COPY_MIX; /* both sides copy a part of chunks */
```